### PR TITLE
Expose sharding configuration via telemetry

### PR DESCRIPTION
**Pull request by Mikulas**
**What this PR does / why we need it**:

Kube-state-metrics component has sharding capabilities, but the sharding configuration is hidden in command arguments and cannot be easily tested after deployment. This may result in multiple forms of misconfiguration, such as:
* part of kubernetes objects may not be processed by any shard (0/3, 1/3, missing 2/3)
* some shards are duplicated, resulting in simple/naive queries yielding incorrect values (0/3, 0/3, 1/3, 2/3)
* variant of the previous state is a residue of old deployment (0/1, 0/3, 1/3, 2/3)

Auto-sharding does not by itself prevent any of those errors, but presumably makes them far less common.

This PR adds two new metrics to kube-state-metrics-self telemetry, namely:
```
# HELP kube_state_metrics_sharding_ordinal Current sharding ordinal/index of this instance
# TYPE kube_state_metrics_sharding_ordinal gauge
kube_state_metrics_sharding_ordinal 0
# HELP kube_state_metrics_sharding_total Number of total shards this instance is aware of
# TYPE kube_state_metrics_sharding_total gauge
kube_state_metrics_sharding_total 1
```

This works with explicit sharding configuration as well as autosharding.

Notes for reviewer:
* I'm not set on `ordinal` being the noun used. The code base uses `shard` which seems less descriptive, but it may be better to keep it consistent?
* Most of the new code is tests. I intentionally did not abstract any of the original tests to keep this separate. It could be eventually refactored, but since this adds new feature it is safer to not also alter the existing tests in the same patch.

### CODE REVIEW

File: pkg/sharding/metrics.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
```
Comment (lilic) [inline pkg/sharding/metrics.go:2]: This needs to be 2021 :)

File: pkg/sharding/metrics.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+// Metrics stores the pointers of kube_state_metrics_sharding_(ordinal|total) metrics.
+type Metrics struct {
+	Ordinal prometheus.Gauge
+	Total   prometheus.Gauge
+}
+
+// NewShardingMetrics takes in a prometheus registry and initializes
+// and registers the kube_state_metrics_list_total and
```
Comment (lilic) [inline pkg/sharding/metrics.go:31]: There was some copy paste leftovers here? :)

File: internal/store/builder.go [hunk 3]
```diff
@@ -74,7 +75,8 @@ func NewBuilder() *Builder {
 
 // WithMetrics sets the metrics property of a Builder.
 func (b *Builder) WithMetrics(r prometheus.Registerer) {
```
Comment (lilic) [inline internal/store/builder.go:77]: Hmm but if we don't use WithSharding this is still created, I would rather make it dependant on the Sharding enabled here.

**Review by lilic — COMMENTED** [review]
Thanks very interesting! Left a few comments

cc @brancz can you have a look as you did the sharding feature, thanks!

File: internal/store/builder.go (hunk 3 above)
Comment (Mikulas) [reply internal/store/builder.go:77]: Fair point! I should have mentioned that in the original description. My understanding is that KSM with no sharding configured behaves identically to `--shard 0 --total-shards 1`. In fact having the metric even for implicit (no-)sharding configuration is one of the benefits.

Is this a reasonable change https://github.com/kubernetes/kube-state-metrics/pull/1344/files#diff-8de36acb4313849c6ece61305374575fa2f18bdb6113d44022ee9749785d4af2R80 to resolve this?

**Comment by Mikulas** [comment]
Hi @brancz, please take a look :)

File: pkg/sharding/metrics.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+// NewShardingMetrics takes in a prometheus registry and initializes
+// and registers the kube_state_metrics_sharding_ordinal and
+// kube_state_metrics_sharding_total metrics. It returns those registered metrics.
+func NewShardingMetrics(r prometheus.Registerer) *Metrics {
+	return &Metrics{
+		Ordinal: promauto.With(r).NewGauge(
+			prometheus.GaugeOpts{
+				Name: "kube_state_metrics_sharding_ordinal",
```
Comment (brancz) [inline pkg/sharding/metrics.go:37]: `kube_state_metrics_shard_ordinal` would be more aligned with naming best practices

File: pkg/sharding/metrics.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+				Help: "Current sharding ordinal/index of this instance",
+			},
+		),
+		Total: promauto.With(r).NewGauge(
+			prometheus.GaugeOpts{
+				Name: "kube_state_metrics_sharding_total",
```
Comment (brancz) [inline pkg/sharding/metrics.go:43]: `kube_state_metrics_total_shards` would be more aligned with naming best practices. `_total` is reserved for counters.

**Review by brancz — COMMENTED** [review]
Super nice! Just a small nit on naming of the metrics, otherwise this is amazing! :100:

File: internal/store/builder.go (hunk 3 above)
Comment (brancz) [reply internal/store/builder.go:77]: Registering the metrics is fine I think, calling `.WithSharding(0, 1)` does seem a bit odd here, just in terms of placement of this call.

Comment (Mikulas) [reply internal/store/builder.go:77]: Ok, I've removed it again. Honestly I can't think of any way that would decouple sharding and metrics without overly complicating the code. The downside of current implementation (without the explicit `WithSharding(0,1)` call) is that if for whatever reason the `WithSharding` is not called later on, `kube_state_metrics_total_shards` will be set to `0` and incorrect. But hopefully this is rather academic; the builder is only used once in this specific file for this component.

**Comment by Mikulas** [comment]
Wonderful, thanks :) I updated the metric names as well as the builder, PTAL

File: internal/store/builder.go (hunk 3 above)
Comment (lilic) [reply internal/store/builder.go:77]: Yes there doesn't seem to be a smooth solution agreed. I am fine with this one, I believe its better than the alternative.

File: README.md [hunk 6]
```diff
@@ -150,15 +150,16 @@ http_request_duration_seconds_sum{handler="metrics",method="get"} 0.021113919999
 http_request_duration_seconds_count{handler="metrics",method="get"} 30
 ```
 
-kube-state-metrics exposes the build info metric:
+kube-state-metrics also exposes build and configuration metrics:
 ```
 kube_state_metrics_build_info{branch="master",goversion="go1.15.3",revision="6c9d775d",version="v2.0.0-beta"} 1
+kube_state_metrics_shard_ordinal 0
```
Comment (lilic) [inline README.md:156]: Do you mind including a couple of use cases for these metrics, similar to how the `kube_state_metrics_build_info ` has below. Thanks!

(So either any alert suggestions or when users should use this)

**Review by lilic — APPROVED** [review]
One more thing left, otherwise this looks good!

File: README.md (hunk 6 above)
Comment (Mikulas) [reply README.md]: Good point, added

**Comment by Mikulas** [comment]
While adding the README documentation I thought of adding example alerts. It proved impossible to construct generic queries without also exposing the `shard_ordinal` as a label due to promql limitations. Without that the query would not be correct when there are multiple replicas for each shard.

I understand this is not a minor change and sorry for disrupting the process but the original proposal was not as useful for alerting as I hoped it to be.

File: jsonnet/kube-state-metrics-mixin/alerts.libsonnet [hunk 7]
```diff
@@ -38,6 +38,43 @@
...
+            // Each shard ordinal is assigned a binary position (2^ordinal) and we compute a sum of those.
+            // This sum is compared to the expected number (2^total_shards - 1).
+            // Result of zero all shards are being scraped, anything else indicates an issue.
+            // A handy side effect of this computation is the result indicates what ordinals are missing.
+            // Eg. a result of "5" decimal, which translates to binary "101", means shards #0 and #2 are not available.
+            expr: |||
+              2^max(kube_state_metrics_total_shards{%(kubeStateMetricsSelector)s}) - 1
+                -
+              sum( 2 ^ max by (shard_ordinal) (kube_state_metrics_shard_ordinal{%(kubeStateMetricsSelector)s}) )
+              != 0
```
Comment (Mikulas) [inline jsonnet/kube-state-metrics-mixin/alerts.libsonnet:67]: Without the `max by (shard_ordinal)` the best we can do is `max by (pod)` but that would only work if there is at most 1 replica for each shard.

File: jsonnet/kube-state-metrics-mixin/alerts.libsonnet [hunk 8]
```diff
@@ -38,6 +38,43 @@
...
+              description: 'kube-state-metrics pods are running with different --total-shards configuration, some Kubernetes objects may be exposed multiple times or not exposed at all.',
+            },
+          },
+          {
+            alert: 'KubeStateMetricsShardsMissing',
+            // Each shard ordinal is assigned a binary position (2^ordinal) and we compute a sum of those.
+            // This sum is compared to the expected number (2^total_shards - 1).
+            // Result of zero all shards are being scraped, anything else indicates an issue.
+            // A handy side effect of this computation is the result indicates what ordinals are missing.
+            // Eg. a result of "5" decimal, which translates to binary "101", means shards #0 and #2 are not available.
```
Comment (Mikulas) [inline jsonnet/kube-state-metrics-mixin/alerts.libsonnet:62]: There is an assumption that there are at most 53 shards, due to the _IEEE 754 double precision_. https://prometheus.io/docs/introduction/faq/#why-are-all-sample-values-64-bit-floats-i-want-integers

**Comment by lilic** [comment]
> I understand this is not a minor change and sorry for disrupting the process but the original proposal was not as useful for alerting as I hoped it to be.

No worries, often this is how we find out which metrics make sense :)

**Review by lilic — COMMENTED** [review]
Thank for the changes, would love to see the unit tests for these rules. From first glance looks okay to me otherwise.

**Comment by Mikulas** [comment]
Thank you @lilic, very good point with the tests. I added promtool based unit tests for the new alerts and also a CI job, but that one may not be completely functional; I suspect that remains to be seen after a merge.

**Comment by lilic** [comment]
Strange the CI is not reporting back, do you mind just force pushing to retrigger them? Thanks @Mikulas !

**Comment by Mikulas** [comment]
Ah yes I broke CI by adding the new alert job and it never reported back. Fixed :)

**Pull request merged**